### PR TITLE
Release: enforce one-change forward-port closeout

### PR DESCRIPTION
**Pull request by justin808**
## Why

The React on Rails 17.0.0 close-out exposed a process gap: release commits that
were not yet on `main` were collected into one large forward-port PR (#4734),
while already-landed changes, release-only metadata, and changelog authority
were mixed into the same review surface.

This PR makes the intended policy executable:

- squash changelog/release PRs;
- forward-port each missing non-release source change in its own PR;
- skip changes already represented on `main`;
- keep the final changelog as a separately acknowledged authority;
- fail closed when source provenance is ambiguous;
- make release-branch deletion atomic, recoverable, and race-aware.

## Implementation

- Reworks `script/release-forward-port` around one-source-change planning,
  provenance/equivalence checks, `--only`, `--all`, `--check`,
  `--ack-manual`, and authoritative changelog handling.
- Strengthens `script/release-finish` with pre-delete durability checks,
  atomic recovery-ref creation, source/main race detection, leased cleanup,
  and automatic restoration when final verification fails.
- Adds exhaustive Ruby and shell regression coverage for provenance,
  incomplete changelogs, branch recreation, main advancement, stale leases,
  recovery cleanup, and operator confirmation boundaries.
- Updates the release-train runbook and agent policy to document the
  one-change-at-a-time close-out contract.

## Process-gap disposition

- **Mechanism target:** `script`
- **Motivating miss:** combined PR #4734 obscured which release commits were
  already on `main`, which needed individual review, and which were release
  metadata only.
- **Replay evidence:** the corrected flow produced the separate changelog PR
  #4742 and source-change PRs #4781 and #4782, while recognizing #4668 as the
  already-merged equivalent of release source #4660. The live
  `release/17.0.0` check reports 0 picks, 34 skips, and 2 explicit manual
  acknowledgements.
- **Non-goal:** this PR does not forward-port product code, restamp the
  17.0.0 changelog, or change released package versions.

## Verification

- `bash script/release-forward-port-test.bash` — 26/26 passed
- `bash script/release-finish-test.bash` — 36/36 passed
- `bundle exec rspec spec/react_on_rails/release_forward_port_script_spec.rb`
  from `react_on_rails/` — 109 examples, 0 failures
- OSS RuboCop — 243 files inspected, 0 offenses; targeted RuboCop also passed
  after the final provenance regression edit
- Ruby syntax, Bash syntax, ShellCheck, Prettier, diff checks, and
  `script/check-docs-sidebar` passed
- Live `release/17.0.0` code completeness check passed with the two documented
  manual acknowledgements
- Live authoritative changelog check passed against source commit
  `44d25171e7fc2eb86d7ccba622862b2a371aae60`

## Review receipt

- The primary pre-push review found a P1 false provenance-skip case involving
  incidental PR numbers in source titles; the implementation and two distinct-
  identity regressions were added before this final candidate.
- The primary reviewer then degraded by recursively launching itself during the
  final-verdict rerun.
- Independent exact-head review at
  `1eb3cbde4e0ebf9da88a06828479218079c15e58` returned `MERGE_READY` with no
  P0-P3 findings.
- The final simplify gate timed out without source edits; its temporary
  comparison artifacts were removed and the reviewed head remained unchanged.
- Strict untrusted-input preflight passed; the only suspicious-text heuristic
  was acknowledged after confirming it was ordinary changelog prose.

## Changelog

`not_user_visible` — internal maintainer tooling, tests, and contributor docs.

This tooling/process PR should be squash merged.

## Summary by CodeRabbit

* **New Features**
  * Added plan-check mode for forward-port operations with scoped selection, explicit acknowledgements for inspected manual items, and stronger changelog reconciliation validation.
  * Extended release close-out to require forward-port and changelog completion before proceeding.
* **Bug Fixes**
  * Blocked release close-out during active cherry-picks and until all required code/changelog forward-port work is finished.
  * Improved safety for ephemeral release-branch deletion with atomic, lease-based race handling and recovery.
* **Documentation**
  * Updated release automation runbooks and close-out guidance with safer sequencing and queue-based forward-port workflows.
* **Tests**
  * Expanded coverage for CLI edge cases, changelog reconciliation completion/deduping, and race-condition close-out behavior.

### CODE REVIEW

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

File: internal/contributor-info/release-train-runbook.md [hunk 1]
```diff
@@ -288,29 +288,105 @@ dependency. For `17.0.0` and `react-on-rails-rsc@19.2.1`, use this order:
...
+  --target release/reconcile-17.0.0-changelog \
+  --changelog \
+  --dry-run
+script/release-forward-port \
+  --source origin/release/17.0.0 \
+  --target release/reconcile-17.0.0-changelog \
+  --changelog
+git diff --check
+git add CHANGELOG.md
+git commit -m "Reconcile X.Y.Z release changelog on main"
```
[the remark this answers is not in the document]
Comment (justin808) [reply internal/contributor-info/release-train-runbook.md]: Fixed in c4de4b2. The copy-pasteable 17.0.0 reconciliation command now commits with `Reconcile 17.0.0 release changelog on main`.

File: script/release-forward-port [hunk 2]
```diff
@@ -545,11 +976,12 @@ class ReleaseForwardPort
     end.uniq
   end
 
-  def patch_equivalent_commit_present_on_target?(sha, target)
-    source_patch_id = patch_id_for_commit(sha)
+  def patch_equivalent_commit_present_on_target?(sha, target, candidates: nil, zero_context: false)
+    source_paths = changed_paths_for_commit(sha) - [CHANGELOG_FILE]
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Independently verified and fixed in c4de4b2. Path-scoped equivalence now excludes the generated LLM artifacts alongside `CHANGELOG.md`, and mixed code+artifact commits get the path-history fallback even when `git cherry` sees different generated output. The new regression proves identical code with divergent generated text is SKIP, not PICK.

File: script/release-finish [hunk 3]
```diff
@@ -274,63 +302,228 @@ class ReleaseFinish
...
+      remote,
+      "#{checked_source_sha}:#{recovery_ref}",
+      ":#{source_ref}"
+    ]
+  end
+
+  def verify_main_after_delete!(remote, release_branch, checked_main_sha:, checked_source_sha:, recovery_ref:)
+    fetch_remote!(remote)
+    current_main_sha = git("rev-parse", "#{remote}/main").strip
+    recreated_source_sha = remote_branch_sha(remote, release_branch)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-finish]: Fixed in c4de4b2. Post-delete fetch/main/source verification I/O is now wrapped in a fail-closed `AbortError` that names the recovery branch, says verification did not complete, and avoids claiming current remote state. A real throwaway-remote regression forces `ls-remote` to fail after deletion and proves the checked source remains on recovery.

File: script/release-finish [hunk 4]
```diff
@@ -274,63 +302,228 @@ class ReleaseFinish
...
+    raise GitError, "could not launch #{FORWARD_PORT_SCRIPT}" if launched.nil? || status.nil?
 
-    raise AbortError,
-          "#{FORWARD_PORT_SCRIPT} did not complete (conflict or MANUAL item). Resolve per the runbook, " \
-          "then re-run close-out; the helper is idempotent and the branch deletion below was not run."
+    return true if status.success?
+    return false if status.exitstatus == FORWARD_PORT_INCOMPLETE_STATUS
+
+    raise GitError, "#{FORWARD_PORT_SCRIPT} verification failed with status #{status.exitstatus}"
   end
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-finish:368]: Verified and declined: close-out resolves `origin/release/X.Y.Z` with `git rev-parse` before invoking the helper, so an invalid source never becomes ordinary pending work. When `release-forward-port --check` is invoked directly, Git errors already exit with status 3 and `run_forward_port_check` treats every non-0/non-1 status as a verifier failure. Changing normal-mode exit semantics is unnecessary for this path.

File: script/release-finish-test.bash [hunk 5]
```diff
@@ -4,10 +4,10 @@
...
 #
 # Every case runs against a throwaway `mktemp -d` git repo and exercises only the
-# dry-run output and the guard/abort paths. NOTHING here runs a real release,
-# branch deletion, or push: promote stops at the rake-release confirmation under
-# dry-run, and close-out only invokes the forward-port DRY-RUN plan plus printed
-# branch-deletion. The harness never adds a network remote.
+# dry-run output and the guard/abort paths. NOTHING here runs a real network
+# release or push: promote stops at the rake-release confirmation under dry-run.
+# Close-out tests use only a throwaway local bare remote when exercising the
+# final branch-deletion gate.
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-finish-test.bash:11]: Independently verified and fixed in c4de4b2. The release-finish shell harness now exports `GIT_CONFIG_GLOBAL=/dev/null` and `GIT_CONFIG_NOSYSTEM=1`, matching the sibling harness so signing, hooks, and system templates cannot contaminate throwaway repositories.

File: script/release-finish-test.bash [hunk 6]
```diff
@@ -435,38 +443,111 @@ test_close_out_yes_non_dry_run_deletes_branch_when_forward_port_pushed() {
...
-  # The forward-port DID apply locally...
-  assert_contains "$RF_OUT" "Forward-port complete" "close-out local-only applied the pick"
-  # ...but the gate refused the delete because main was not pushed.
-  assert_contains "$RF_OUT" "not yet on origin/main" "close-out local-only message"
-  assert_not_contains "$RF_OUT" "git push origin --delete" "close-out local-only must not delete"
+  assert_status 1 "$RF_STATUS" "close-out incomplete status"
+  assert_contains "$RF_OUT" "release close-out is not complete on origin/main" "close-out incomplete message"
+  assert_contains "$RF_OUT" "each remaining source change in its own PR" "close-out incomplete next step"
+  assert_not_contains "$RF_OUT" "Forward-port complete" "close-out must not apply picks directly"
+  assert_not_contains "$RF_OUT" "git push origin --delete" "close-out incomplete must not delete"
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-finish-test.bash]: Fixed in c4de4b2. The code-incomplete guard now asserts that the actual atomic deletion command (`git push --atomic`) is absent.

File: script/release-finish-test.bash [hunk 7]
```diff
@@ -435,38 +443,111 @@ test_close_out_yes_non_dry_run_deletes_branch_when_forward_port_pushed() {
...
+  git fetch -q origin
+
+  run_rf close-out 1.0.0 --yes
+
+  assert_status 1 "$RF_STATUS" "close-out changelog-incomplete status"
+  assert_contains "$RF_OUT" "CHECK PASSED: the release source has no remaining code forward-port work" \
+    "close-out changelog-incomplete code check"
+  assert_contains "$RF_OUT" "CHECK FAILED: release changelog reconciliation still needs its own squash PR" \
+    "close-out changelog-incomplete changelog check"
+  assert_not_contains "$RF_OUT" "git push origin --delete" "close-out changelog-incomplete must not delete"
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-finish-test.bash]: Fixed in c4de4b2. The changelog-incomplete guard now asserts that the actual atomic deletion command (`git push --atomic`) is absent.

File: script/release-forward-port [hunk 8]
```diff
@@ -245,23 +387,141 @@ class ReleaseForwardPort
     return active_source_revert_skip_reason(subject, target, target_version) if
       source_standard_revert_on_target?(sha, target)
 
+    return "source commit identifies a live upstream PR already present on #{target}" if
+      source_narrated_pr_origin_on_target?(sha, target)
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed in `406cc3212dfa1ffdb3e04f823b32f1493aa8c850`.

Generic `Backports #N` narration no longer silently skips a release commit. It falls through to patch/source-bound proof and, when only the named upstream PR is live, becomes `MANUAL` with an explicit inspection requirement. Explicit whole-commit equivalence and composite/replacement provenance now also require every non-changelog source path to be covered by live target commits.

Regression coverage proves both generic narration and an explicit equivalence sentence cannot hide an uncovered release-only path. The focused cases, full forward-port RSpec suite (112 examples), forward-port shell suite (26 tests), closeout shell suite (38 tests), RuboCop, syntax, ShellCheck, Prettier, docs sidebar, and live 17.0.0 code plan all pass. The live plan now has zero `PICK` items and surfaces the three adapted source-backed commits as inspected `MANUAL` acknowledgements instead of silent skips.

**Comment by justin808** [comment]
Current-head review fix: `406cc3212dfa1ffdb3e04f823b32f1493aa8c850`

- Fixed the Codex P1 provenance gap: generic upstream-PR narration is now fail-closed `MANUAL`, while stronger skip forms require source-path coverage by live target commits.
- Added regressions for a generic backport and an explicit equivalence sentence carrying an uncovered release-only file.
- Updated the runbook fetch command to `git fetch --prune origin main`.

Exact validation:

- forward-port RSpec: 112 examples, 0 failures
- forward-port changelog shell suite: 26/26
- release-finish shell suite: 38/38
- focused current-head provenance/revert cases: 6 examples, 0 failures
- targeted RuboCop: clean
- Ruby/Bash syntax, ShellCheck, Prettier, docs sidebar, `git diff --check`: clean
- pre-commit and pre-push hooks: green
- live `origin/release/17.0.0` → `origin/main` code plan: zero `PICK`; five inspected `ACK-MANUAL` items

The separate changelog checker additionally found five final-release entries still absent from `main`; those will be handled in their own squash changelog PR, not folded into this tooling PR.

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

File: script/release-finish [hunk 9]
```diff
@@ -274,63 +302,247 @@ class ReleaseFinish
...
-    forwarded = system(FORWARD_PORT_SCRIPT, "--source", source, "--target", "main")
-    return if forwarded
+    puts "\nNext: follow the runbook's one-source-PR-at-a-time loop, then merge the release/changelog squash PR."
+    false
+  end
+
+  def run_forward_port_check(*)
+    launched = system(FORWARD_PORT_SCRIPT, *)
+    status = Process.last_status
+    raise GitError, "could not launch #{FORWARD_PORT_SCRIPT}" if launched.nil? || status.nil?
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-finish:362]: Verified and declined: Ruby supports anonymous rest-argument forwarding in this form, and the repository requires Ruby 3.3+. Both `ruby -c script/release-finish` and the complete release-finish shell suite pass (38/38), including calls through `run_forward_port_check`. No syntax or runtime defect is present.

File: script/release-forward-port [hunk 10]
```diff
@@ -1352,23 +2058,146 @@ class ChangelogReconciler
...
+  end
+
+  def retained_target_entries(target, removed_sections, replacement_prs)
+    target[:sections]
+      .reject { |section| removed_sections.include?(section) }
+      .flat_map do |section|
+        entries = section_entries(section)
+        next entries unless section[:version] == UNRELEASED
+
+        entries.reject { |entry| replacement_prs.include?(entry.pr_number) }
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed in `d2207322f2f2cd43a16772fa6a5ee7756e789b5d`.

When the source has an active final stable section but no source-bound authoritative changelog commit (or explicit source acknowledgement), that target stable section is now excluded from ordinary retained-history de-duplication. The reconciliation therefore remains incomplete instead of allowing an unbound final section to make `--changelog --check` pass.

The new regression copies the final section verbatim to the target without the source-SHA sentence, proves check mode exits 1, and then proves the exact source acknowledgement makes the unchanged authoritative section pass. Full forward-port RSpec is green at 113 examples.

File: script/release-forward-port [hunk 11]
```diff
@@ -421,6 +782,131 @@ class ReleaseForwardPort
...
+        target_commit_live_on_target?(target_sha, target)
+      end
+    end.uniq
+  end
+
+  def source_paths_covered_by_target_commits?(source_sha, target_shas)
+    source_paths = changed_paths_for_commit(source_sha) - [CHANGELOG_FILE] - GENERATED_LLM_FILES
+    target_paths = target_shas.flat_map { |target_sha| changed_paths_for_commit(target_sha) }.uniq
+
+    source_paths.any? && (source_paths - target_paths).empty?
```
[the remark this answers is not in the document]
Comment (justin808) [reply script/release-forward-port]: Fixed in `d2207322f2f2cd43a16772fa6a5ee7756e789b5d`.

I removed narration-only automatic skips entirely instead of attempting a fuzzy content heuristic. Source-commit statements naming upstream PRs—including equivalence, replacement, and composite forms—now produce `MANUAL` only after every required target commit is live. Automatic skips require patch equivalence, `cherry-pick -x`, or exact source-bound provenance recorded by the target forward-port commit.

This eliminates both single- and multi-PR path-overlap false skips. The live 17.0.0 plan remains at zero `PICK` items and now exposes seven inspected `ACK-MANUAL` items. Full RSpec (113 examples), both shell suites (26/26 and 38/38), RuboCop, syntax, ShellCheck, formatting, docs, and hooks are green.

**Comment by justin808** [comment]
Final review-fix candidate: `d2207322f2f2cd43a16772fa6a5ee7756e789b5d`

- Removed every narration-only automatic code skip. Adapted upstream-PR relationships now require `MANUAL` operator inspection; automatic skips require patch, `-x`, or target-recorded exact source provenance.
- Prevented an unbound active final changelog section from satisfying ordinary retained-history de-duplication.
- Kept the live 17.0.0 code plan at zero `PICK` items with seven explicit inspected acknowledgements.

Exact validation:

- forward-port RSpec: 113 examples, 0 failures
- forward-port changelog shell suite: 26/26
- release-finish shell suite: 38/38
- live code check: zero `PICK`, 29 automatic skips, 7 `ACK-MANUAL`
- targeted RuboCop, Ruby/Bash syntax, ShellCheck, Prettier, docs sidebar, `git diff --check`: clean
- pre-commit and pre-push hooks: green

The Copilot anonymous-forwarding report was independently disproved by Ruby syntax validation and the exercised 38-test release-finish suite.

**Comment by justin808** [comment]
+ci-status

**Comment by justin808** [comment]
+ci-run-hosted

**Comment by justin808** [comment]
@codex review

**Comment by justin808** [comment]
## Address-review summary

Scan scope: full PR history through exact head `d2207322f2f2cd43a16772fa6a5ee7756e789b5d`.

### Mattered
- Addressed the false-skip provenance defect: generic upstream narration became fail-closed `MANUAL`, then all narration-only automatic skips were removed after Claude identified that shared path names were not content proof.
- Addressed the unbound-final-changelog defect: an active final section can no longer satisfy retained-history de-duplication without source-SHA authority or an explicit exact-source acknowledgement.
- Addressed closeout safety/portability findings: isolated throwaway git config, strengthened recovery/error guidance and race tests, ignored regenerated LLM artifacts only during code-patch matching, and guarded every fixture setup step.
- Verified the final exact head with 113 RSpec examples, 26 forward-port shell tests, 38 release-finish shell tests, hosted CI, Claude, CodeRabbit, Copilot, and a final Codex pass reporting no major issues.

### Optional
- Fixed the runbook fetch command to use `git fetch --prune origin main`.
- Declined the additional top-level numbered runbook summary: the section already contains the executable one-change queue and this was presentation-only after the final-candidate debounce point.
- CodeRabbit's ast-grep timeout was tool degradation only; the file passed Ruby parsing, RuboCop, RSpec, and hosted analysis.

### Skipped
- Declined Copilot's anonymous-rest-forwarding syntax report: Ruby 3.3+ supports `def method(*)` / `call(*)`; `ruby -c` and the exercised 38-test closeout suite pass.
- Declined the earlier source-resolution exit-code report: closeout resolves the source before invoking check mode, and check mode already reserves status 1 for incomplete work while treating other nonzero statuses as verifier failures.
- Pure duplicates and reviewer status/boilerplate were safe-to-skip.

Deferred-work tracking: none; no actionable review item remains.

Next default scan starts after this comment. Say `check all reviews` to rescan the full PR.
